### PR TITLE
Parser Refactor and start of Character objects.

### PR DESCRIPTION
**Pull request by viion**
## Todo

- [x] Create **Profile** class
- [x] Create **City** class and attach to profile as a single object
- [x] Create **ClassJob** class and attach to profile as an array of objects
- [x] Create **Collectable** class and attach to profile as an array of objects
- [x] Create **Guardian** class attach to profile as a single object
- [x] Create **Attribute** class attach to profile as collection of attributes
- [x] Create **Gear** class, attach to profile as a single object
  - [x] Create **Item** class attach to "Gear" class
- [x] Create **ActiveClass** class attach to profile as a single object (this will be a ClassJob object)
- [x] Complete hash method

**Additional tasks which could be their own PR:**

- Change validator into a static call, to avoid appending validator to all objects
- Add icon to ClassJob ?
- Add "Total EXP" to ClassJob ? (Obtained from XIVDB)
- Add "EXP Togo" to ClassJob ? (Obtained from XIVDB)
- Add percentages for EXP to ClassJob ?
- Handle ID for Grand Company rank (Difficult due to 3 ranks + male/female titles)
- Provide Name/Icon to Free Company instead of just the ID ? (Could be a SimpleFreeCompany object)

## PR Description
The method 'getCharacterHash()' has been removed, a hash will always be generated. This reduces the code in the API call.

Fixed an issue in toArray() when an array type exists, it needs to loop through the values and all toArray on each one (An array is just an array of other objects)

Moved parse code into traits.

Moved city/gc/guardian back a folder, this folder might get quite long but most of them are single instances (collectable, classjob, item, etc)

Set character id for profile during construct, make it a requirement.

Added "initialize" to parse helper which does the usual: ensureHtml, trim html, setInitialDocument. Nearly all parse methods does this proceedure.

### CODE REVIEW

**Comment by shentschel** [comment]
Looks good :) A Proposal of my side would be to create an entity class 'Character' which contains Profile and the other stuff. That would allow us to return a single class instead of an array. What do you think about this @viion ?

**Comment by viion** [comment]
I was thinking this too, I might make an option, or by default and then have "toArray" as an option to call against it.

Arrays are easier to store and handle, especially with making REST stuff.

**Comment by viion** [comment]
I wonder if Validators should be made static, this is quite ugly:

```
[33] => Lodestone\Entities\Character\Attribute Object
                (
                    [id] => 42
                    [name] => Water
                    [value] => 0
                    [validator:protected] => Lodestone\Validator\BaseValidator Object
                        (
                            [object] => 0
                            [name] => Attribute Value
                            [errors] => Array
                                (
                                )

                        )

                )
```

**Comment by shentschel** [comment]
I think we can make the validation static :)

IMHO for the REST Stuff we can also extend the entities with [JsonSerializable](http://php.net/manual/de/class.jsonserializable.php). This would allow us to define which fields will be shown in the json-object.

**Comment by viion** [comment]
I'd like to avoid JsonSerializable because it's so dependant on PHP version and for things like Achievements with Date/Time it becomes such a huge pain in the ass. Symfony uses it and for years I've had pain with it.

So for now I think it is safe to maintain the `toArray` but it's a method you can call after being provided a `Character` object, eg:

```php
$character = $api->getCharacter(xxx);
echo 'Hello: '. $character->profile->name .' Your hash is: '. $character->hash;

$json = json_encode($character->toArray());
```

**Comment by viion** [comment]
Completed all data appending. I've also improved how `ActiveClass` is handled, since I parse the items anyway, I just need the item `category` appended during that. Then setting the `Active Class` requires no more additional parsing as it takes the `mainhand` and the current recorded class jobs to handle it all. Big win!

Just going to clean up some minor bugs and complete the hash method.

**Comment by viion** [comment]
Looking at memory, I may have introduced a leak, possible. 

There is an obvious huge increase as I found a bug with item data not being provided to the API so it was missing, however due to there being thousands of items it causes a jump from 1.7mb to 13.8mb. One way to reduce this would be for me to provide IDs that are relevant (gear, materia, dyes) from XIVDB, but that is out the scope of the API so I will do that independently.

After this it parses and it goes from `13,805,816` to `14,147,080` (avg), it's a small increment of 0.35mb but it's consistently growing. So not sure.

**Comment by viion** [comment]
Also due to having such a big array, it makes getting the real in-game item id a costly procedure... Defaintly need to refactor

```
Duration: 0.14906600   + 0.00974500  !    Mem: 14,034,768       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.17060100   + 0.02153500  !    Mem: 14,035,696       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.17943300   + 0.00883200  !    Mem: 14,042,328       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.20025600   + 0.02082300  !    Mem: 14,042,384       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.20809800   + 0.00784200  !    Mem: 14,042,376       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.22755600   + 0.01945800  !    Mem: 14,042,432       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.23520200   + 0.00764600  !    Mem: 14,043,184       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.25547100   + 0.02026900  !    Mem: 14,043,240       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.26424500   + 0.00877400  !    Mem: 14,043,456       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.28420100   + 0.01995600  !    Mem: 14,043,512       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.29103200   + 0.00683100  !    Mem: 14,043,792       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.31298700   + 0.02195500  !    Mem: 14,043,848       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.32180700   + 0.00882000  !    Mem: 14,044,496       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.34265600   + 0.02084900  !    Mem: 14,044,552       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.35032500   + 0.00766900  !    Mem: 14,045,016       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.37321700   + 0.02289200  !    Mem: 14,045,072       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.37978800   + 0.00657100  !    Mem: 14,045,360       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.40158400   + 0.02179600  !    Mem: 14,045,416       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.40907500   + 0.00749100  !    Mem: 14,046,528       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.43402500   + 0.02495000  !    Mem: 14,046,584       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.44312200   + 0.00909700  !    Mem: 14,047,048       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.47789100   + 0.03476900  !    Mem: 14,047,104       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.48723800   + 0.00934700  !    Mem: 14,047,528       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.52085300   + 0.03361500  !    Mem: 14,047,584       Line 178   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.53120200   + 0.01034900  !    Mem: 14,047,560       Line 173   in  Lodestone\Modules\XIVDB::getItemId
Duration: 0.55217200   + 0.02097000  !    Mem: 14,047,616       Line 178   in  Lodestone\Modules\XIVDB::getItemId
```

(jump from 0.14 to 0.5s!)

**Comment by viion** [comment]
Things are starting to look better, accessing gear data is the main bottleneck right now that I need to figure out.

screen shot 2017-06-02 at 09 54 15

**Comment by shentschel** [comment]
Maybe we can provide the gear data as another entry point or as an option. Then we would save time for the character parsing:) my 2 cents

**Comment by viion** [comment]
Got things down to something reasonable and matching what I run on XIVSync. Item ID's are still a big sink, (0.1 delay). 

I am considering at this point to make all XIVDB ID conversions an "Option" that you must allow prior to searching. By default it won't convert anything to XIVDB Data ID's. This will make it more beneficial if you wish to just use what the API provides rather than being able to extend onto XIVDB's API.

```
Duration: 0.09682100   + 0.00024400       Mem: 1,919,888        Line 20    in  Lodestone\Parser\Character\TraitGear::parseEquipGear
Duration: 0.10680400   + 0.00998300  !    Mem: 1,947,712        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite War Axe)
Duration: 0.10695700   + 0.00015300       Mem: 1,953,096        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite War Axe)
Duration: 0.11596100   + 0.00900400  !    Mem: 1,953,384        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Sallet of Fending)
Duration: 0.11612100   + 0.00016000       Mem: 1,959,008        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Sallet of Fending)
Duration: 0.12348400   + 0.00736300  !    Mem: 1,956,176        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Hauberk of Fending)
Duration: 0.12361900   + 0.00013500       Mem: 1,961,320        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Hauberk of Fending)
Duration: 0.13061500   + 0.00699600  !    Mem: 1,958,616        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Gauntlets of Fending)
Duration: 0.13093000   + 0.00031500       Mem: 1,963,680        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Gauntlets of Fending)
Duration: 0.13912100   + 0.00819100  !    Mem: 1,960,520        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Tassets of Fending)
Duration: 0.13927400   + 0.00015300       Mem: 1,965,584        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Tassets of Fending)
Duration: 0.14558200   + 0.00630800  !    Mem: 1,962,488        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Archaeoskin Breeches of Fending)
Duration: 0.14569700   + 0.00011500       Mem: 1,967,752        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Archaeoskin Breeches of Fending)
Duration: 0.15185500   + 0.00615800  !    Mem: 1,964,824        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Sabatons of Fending)
Duration: 0.15195500   + 0.00010000       Mem: 1,969,888        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Sabatons of Fending)
Duration: 0.15844800   + 0.00649300  !    Mem: 1,966,976        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Earrings of Fending)
Duration: 0.15857200   + 0.00012400       Mem: 1,971,960        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Earrings of Fending)
Duration: 0.16521400   + 0.00664200  !    Mem: 1,968,952        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Rainbow Ribbon of Fending)
Duration: 0.16539800   + 0.00018400       Mem: 1,972,176        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Rainbow Ribbon of Fending)
Duration: 0.17275100   + 0.00735300  !    Mem: 1,971,432        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Bangle of Fending)
Duration: 0.17290300   + 0.00015200       Mem: 1,974,816        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Mythrite Bangle of Fending)
Duration: 0.17931200   + 0.00640900  !    Mem: 1,973,584        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Agate Ring of Fending)
Duration: 0.17942700   + 0.00011500       Mem: 1,978,888        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Agate Ring of Fending)
Duration: 0.18624800   + 0.00682100  !    Mem: 1,975,696        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Agate Ring of Fending)
Duration: 0.18635200   + 0.00010400       Mem: 1,980,184        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Agate Ring of Fending)
Duration: 0.19255600   + 0.00620400  !    Mem: 1,975,728        Line 66    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Soul of the Warrior)
Duration: 0.19268400   + 0.00012800       Mem: 1,980,632        Line 72    in  Lodestone\Modules\XIVDBDataTrait::getItemId(Soul of the Warrior)
Duration: 0.19747500   + 0.00479100  !    Mem: 1,973,168        Line 120   in  Lodestone\Parser\Character\TraitGear::parseEquipGear
```

**Comment by shentschel** [comment]
Is the items array that big? :o

Von meinem iPhone gesendet

**Comment by viion** [comment]
Added static options

```php
$api->xivdb->setOptions([
    'items' => false,
    'exp' => true,
    'classjobs' => true,
    'profile' => true,
    'collectables' => true,
    'attributes' => true,
]);
```

Having most off is fine, and there isn't much of a time difference for most. Disabling "classjobs" will mean "ActiveClass" is pretty much useless and won't convert to Job as I use XIVDB data for that to avoid parsing and keep JobClass objects consistent.

For items it adds about 0.03-0.04 delay, most of time delay is in the Parsing of gear. Need to look into it.

**OFF**
```
Duration: 0.09628000   + 0.00023600       Mem: 1,919,808        Line 20    in  Lodestone\Parser\Character\TraitGear::parseEquipGear
Duration: 0.18613100   + 0.08985100  !    Mem: 1,952,448        Line 120   in  Lodestone\Parser\Character\TraitGear::parseEquipGear
```

**ON**
```
Duration: 0.09326200   + 0.00026000       Mem: 1,919,808        Line 20    in  Lodestone\Parser\Character\TraitGear::parseEquipGear
Duration: 0.19745100   + 0.00489300  !    Mem: 1,973,088        Line 120   in  Lodestone\Parser\Character\TraitGear::parseEquipGear
```

need to investigate why parsing gear is so slow all together.

**Comment by viion** [comment]
> @shentschel Is the items array that big? :o

Unfortunately, yes, I've made a custom route on XIVDB specifically for the lodestone library: http://api.xivdb.com/lodestone/php/items

There are 10137 (out of 17,000) possible items. These include: Dyes, Materia, Soul Crystals, Weapons, Armor and Shields.

It's about a 10mb array if fully loaded so the script now breaks it up by 2 characters of md5, this results in 256 files being created (2 digit hex) and makes the whole thing super light (hence memory reduction from 14mb down to 2mb again!)

However, even with all that, its now lightning fast but the parse itself is slow... :( Need to find the main cause.

**Comment by shentschel** [comment]
Hmm do you load the Json data from disk? If yes, maybe the I/O speed is slow so that PHP needs most of the time for loading the file.

**Comment by viion** [comment]
I/O speed is not bad and thankfully doesn't cost resources, I'd rather load a file quickly than have a 14mb array in memory which stacks up.

The main issue with gear is not the json, with that removed you only reduce speed by 0.01, the main problem is in the gear parsing logic itself, so maybe there is a leak somewhere

**Comment by shentschel** [comment]
Hmm I can have a look at it on Monday. Currently I have a little connection problem (my mum in law has zero internet in her village)

**Comment by viion** [comment]
Been mega busy, only just now had chance to get back at this. Looking into `Lodestone\Parser\Character\TraitGear::parseEquipGear` to see why it causes so much

Base: Average 111.25 ms

**Test 1:** Remove parsing of: mirage, creator, dye, materia
- Result: Average 85.2 ms

**Test 2:** Remove category
- Result: Average 30.83

Some how the category code is responsible for a 0.05s increase, this is something I added recently.

```
// category
$category = $box->find('.db-tooltip__item__category', 0)->plaintext;
$category = explode("'", $category)[0];
$category = str_ireplace(['Two-handed', 'One-handed'], null, $category);
$item->setCategory(trim($category));
```

**Comment by viion** [comment]
Doh figured it out, it's using `$box` instead of `$node` so it's scanning the entire doc tree each time. It's also only needed on the first interval

Average after: 31.5

File: src/Lodestone/Parser/Character/TraitGear.php [hunk 1]
```diff
@@ -119,5 +131,11 @@ protected function parseEquipGear()
 
         unset($box);
         Benchmark::finish(__METHOD__,__LINE__);
+
+        $b = round(microtime(true) * 1000) - $a;
+        file_put_contents(__DIR__.'/gear.log', $b . "\n", FILE_APPEND);
```
Comment (shentschel) [inline src/Lodestone/Parser/Character/TraitGear.php]: Is this logging method for gear only? :D

**Comment by shentschel** [comment]
Looks better :) So which things need to be done next?

File: src/Lodestone/Parser/Character/TraitGear.php (hunk 1 above)
Comment (viion) [inline src/Lodestone/Parser/Character/TraitGear.php]: Yes this is logging just gear so I was able to debug it, i wanted an exact figure just for gear alone.

**Comment by viion** [comment]
I want to try improve it further, I may remove all CSS selectors from gear parsing and use a `explode` break method instead, I still feel `0.03` to get very minimal information is quite high.

**Comment by viion** [comment]
Been investing this **a lot** yesterday, I normally bring my laptop home from work over weekends but since I'm off next week for SB I decided not to and work on this on my laptop.

Suddenly, I go from 0.18s ms parse to 0.4-0.5 and I'm thinking, **what just happened...*

Anyway, my work laptop is a lot more powerful than my home laptop, so I figured i would test the script on a live XIVSync server to get a true result.

And woah!

```
[2017-06-10 17:31:03-.545][Lodestone\Parser\Character\Parser][67] PARSE DURATION: 84 ms
```

Consistently around 80ms (0.08) So this is below the 0.1 goal. My Sync servers are on extremely powerful CPU's (cloud computing and all) which has a big factor on this.

To compare it to the old API before we did any classes/objects, the range is about `0.05 ms` to `0.15 ms` Some going as high as `0.25`, I think it depends on the character.

The Lodestone ID: **6133001** parsed in `268 ms` (0.268), on average the new code is around `115` ms to `250` ms, so it's close but getting better (i think this char might have all mounts/minions, or a lot).

I am thinking of modifying profile information to a `array-explode` logic, this method is very hacky as it's expecting array index's to be specific, however it is faster as there is no string searching or DOMDocument/Node interaction... I've done it for most of the gear except materia (can be 1-5 materia slots). I may do it just to see the if the results are worth it.

Right now the changes are in a very good stage.

**Comment by shentschel** [comment]
Sounds like we got a solid system even with OOP :) And that allows us to be more efficient while maintaining the code :D 

I think we earned us a good cold cola :D *high five*

**Comment by viion** [comment]
I've now converted all of profile stuff to an Array-Explode filtering method instead of using DomDocument. This is a bit more "hacky" and "hard coded" but it is incredibly fast and hard to give up. I think for Characters it's needed because of the expected high-parse rate, where as pages such as Lodestone Topics a DomDocument is still better.

The avg parse time has gone from 38-45ms to 6-8ms

Previous:
```
Duration: 0000000000   + 0000000000       Mem: 2,553,328        Line 54    in  Lodestone\Parser\Character\Parser::parse
Duration: 0.00014400   + 0.00014300       Mem: 2,554,496        Line 69    in  Lodestone\Parser\Character\TraitProfile::parseProfileBasic
Duration: 0.00525000   + 0.00510600  !    Mem: 2,557,368        Line 92    in  Lodestone\Parser\Character\TraitProfile::parseProfileBasic
Duration: 0.00537700   + 0.00012700       Mem: 2,556,640        Line 100   in  Lodestone\Parser\Character\TraitProfile::parseProfileBiography
Duration: 0.00754100   + 0.00216400  !    Mem: 2,557,496        Line 109   in  Lodestone\Parser\Character\TraitProfile::parseProfileBiography
Duration: 0.01030000   + 0.00275900  !    Mem: 2,558,232        Line 119   in  Lodestone\Parser\Character\TraitProfile::parseProfileRaceClanGender
Duration: 0.03388300   + 0.02358300  !    Mem: 3,057,328        Line 133   in  Lodestone\Parser\Character\TraitProfile::parseProfileRaceClanGender
Duration: 0.03449200   + 0.00060900       Mem: 3,057,856        Line 143   in  Lodestone\Parser\Character\TraitProfile::parseProfileNameDay
Duration: 0.03493100   + 0.00043900       Mem: 3,058,888        Line 147   in  Lodestone\Parser\Character\TraitProfile::parseProfileNameDay
Duration: 0.03499000   + 0000000000       Mem: 3,059,424        Line 157   in  Lodestone\Parser\Character\TraitProfile::parseProfileGuardian
Duration: 0.03583100   + 0.00084100       Mem: 2,895,256        Line 171   in  Lodestone\Parser\Character\TraitProfile::parseProfileGuardian
Duration: 0.03589300   + 0000000000       Mem: 2,895,752        Line 179   in  Lodestone\Parser\Character\TraitProfile::parseProfileCity
Duration: 0.03654100   + 0.00064800       Mem: 2,903,400        Line 194   in  Lodestone\Parser\Character\TraitProfile::parseProfileCity
Duration: 0.03691800   + 0.00037700       Mem: 2,903,952        Line 204   in  Lodestone\Parser\Character\TraitProfile::parseProfileGrandCompany
Duration: 0.03793100   + 0.00101300       Mem: 2,914,576        Line 222   in  Lodestone\Parser\Character\TraitProfile::parseProfileGrandCompany
Duration: 0.03799800   + 0000000000       Mem: 2,915,008        Line 232   in  Lodestone\Parser\Character\TraitProfile::parseProfileFreeCompany
Duration: 0.03871700   + 0.00071900       Mem: 2,916,392        Line 237   in  Lodestone\Parser\Character\TraitProfile::parseProfileFreeCompany
```

New:
```
Duration: 0000000000   + 0000000000       Mem: 2,551,760        Line 54    in  Lodestone\Parser\Character\Parser::parse
Duration: 0.00010600   + 0.00010400       Mem: 2,552,928        Line 57    in  Lodestone\Parser\Character\TraitProfile::parseProfileBasic
Duration: 0.00093200   + 0.00082600       Mem: 2,555,800        Line 80    in  Lodestone\Parser\Character\TraitProfile::parseProfileBasic
Duration: 0.00109900   + 0.00016700       Mem: 2,555,072        Line 88    in  Lodestone\Parser\Character\TraitProfile::parseProfileBiography
Duration: 0.00402700   + 0.00292800  !    Mem: 2,555,928        Line 97    in  Lodestone\Parser\Character\TraitProfile::parseProfileBiography
Duration: 0.00418100   + 0.00015400       Mem: 2,556,456        Line 107   in  Lodestone\Parser\Character\TraitProfile::parseProfileRaceClanGender
Duration: 0.00481300   + 0.00063200       Mem: 2,557,544        Line 124   in  Lodestone\Parser\Character\TraitProfile::parseProfileRaceClanGender
Duration: 0.00487000   + 0000000000       Mem: 2,557,880        Line 134   in  Lodestone\Parser\Character\TraitProfile::parseProfileNameDay
Duration: 0.00556000   + 0.00069000       Mem: 2,559,336        Line 141   in  Lodestone\Parser\Character\TraitProfile::parseProfileNameDay
Duration: 0.00563800   + 0000000000       Mem: 2,559,320        Line 151   in  Lodestone\Parser\Character\TraitProfile::parseProfileGuardian
Duration: 0.00645400   + 0.00081600       Mem: 2,395,608        Line 164   in  Lodestone\Parser\Character\TraitProfile::parseProfileGuardian
Duration: 0.00651400   + 0000000000       Mem: 2,395,472        Line 172   in  Lodestone\Parser\Character\TraitProfile::parseProfileCity
Duration: 0.00696600   + 0.00045200       Mem: 2,403,040        Line 185   in  Lodestone\Parser\Character\TraitProfile::parseProfileCity
Duration: 0.00701600   + 0000000000       Mem: 2,403,576        Line 195   in  Lodestone\Parser\Character\TraitProfile::parseProfileGrandCompany
Duration: 0.00778700   + 0.00077100       Mem: 2,414,488        Line 216   in  Lodestone\Parser\Character\TraitProfile::parseProfileGrandCompany
Duration: 0.00783700   + 0000000000       Mem: 2,414,304        Line 226   in  Lodestone\Parser\Character\TraitProfile::parseProfileFreeCompany
Duration: 0.00854200   + 0.00070500       Mem: 2,416,016        Line 237   in  Lodestone\Parser\Character\TraitProfile::parseProfileFreeCompany
```


between parseProfileBiography & parseProfileRaceClanGender was an expensive call:

```
$box = $this->getDocumentFromRange('character__profile__data__detail', 'character__view');
```

There were a couple other expensive calls, such as `$this->getDocumentFromRangeCustom(40,100)`

**Comment by viion** [comment]
I've removed `hash` (which calls `toArray`) from being an automatic calculated parameter. It can adds a significant delay (eg: `0.1094` to `0.1246`. This kind of call would suit better being optional, so now it can be called via: 
```
$api->hash->hashCharacter($data)
```

This would allow it to be consistent with other forms, eg: `hashLinkshell` and `hashFreeCompany`.

The API call will now return a `CharacterProfile` instead of just `Profile`, this is so we can have:
```
CharacterProfile
CharacterAchievements
CharacterFriends
LinkshellMembers
FreeCompanyProfile
FreeCompanyMembers
```

etc, very consistent classes.

**Comment by viion** [comment]
All tasks for this have been PR have completed, right now it just needs a bit of a review and any code cleans.

Due to code structure changes, other features will need moving, for example parsing code is now in `Parser/Character/X`, so respectively, for other features, their folders would be: `Parser/FreeCompany/X`, `Parser/Linkshell/X` etc. 

Considering moving "Achievement" related stuff under the "Character" feature.

**Comment by shentschel** [comment]
I'll look into it tomorrow :)

**Comment by viion** [comment]
Merged in PR and moved Hash method to the entity, I think all your changes were really good.

**Materia Parse**
I had a go at Materia but couldn't improve it. I'm half thinking to leave it for now as a side-product as the time cost is not too bad and I think we should move onto other stuff. The next big job is getting all other features (FC Parse, Linkshell Parse, Achievements) etc into the same fashion as we've done on Profile. We could make a note to revisit.

I will probably merge this before we move onto other stuff. If you're happy with it all.

**Comment by shentschel** [comment]
I'm happy with all stuff :) I think we did a solid-job 👍

**Pull request merged**